### PR TITLE
[GLUTEN-810][VL] Support hdfs delegation token

### PR DESCRIPTION
**Pull request by ccat3z**
## What changes were proposed in this pull request?

closes: #810 

* Pass hdfs tokens to native code by session config
* Switch to oap-project/libhdfs3
* Add missing depends for gsasl in vcpkg

## How was this patch tested?

1. Setup hdfs with kerberos
2. Decrease delegation token renew interval to test renewer
   
``` xml
<!-- hdfs-site.xml -->
<property>
  <name>dfs.namenode.delegation.token.max-lifetime</name>
  <!-- 2min -->
  <value>1200000</value>
</property>
<property>
  <name>dfs.namenode.delegation.token.renew-interval</name>
  <!-- 30s -->
  <value>30000</value>
</property>
```

3. Run spark on yarn

``` sh
spark-shell \
  --conf spark.driver.extraClassPath="/path/to/gluten/jar" \
  --conf spark.executor.extraClassPath="/path/to/gluten/jar" \
  --conf spark.plugins=io.glutenproject.GlutenPlugin \
  --conf spark.gluten.sql.columnar.backend.lib=velox \
  --conf spark.shuffle.manager=org.apache.spark.shuffle.sort.ColumnarShuffleManager \
  --conf spark.memory.offHeap.enabled=true \
  --conf spark.memory.offHeap.size=10G \
  --master yarn \
  --keytab /path/to/krb5.keytab \
  --principal "$PRINCIPAL"
```

### CODE REVIEW

**Comment by jinchengchenghh** [comment]
Do you create an issue in facebook velox to discuss about this feature includes updating HiveConnectorConfig? If we can align with them, it will be easy to upstream to facebook velox.

**Comment by tianyouyangying** [comment]
hadoop version:3.3.3 HDFS Router-based Federation
yarn:3.3.3
spark3.3.0
The command is as follows
./spark-sql \
  --conf spark.driver.extraClassPath="/clusteragent/spark_gluten/jars/gluten-thirdparty-lib-centos-7.jar:/clusteragent/spark_gluten/jars/gluten-velox-bundle-spark3.3_2.12-centos_7-0.5.0-SNAPSHOT.jar" \
  --conf spark.executor.extraClassPath="/clusteragent/spark_gluten/jars/gluten-thirdparty-lib-centos-7.jar:/clusteragent/spark_gluten/jars/gluten-velox-bundle-spark3.3_2.12-centos_7-0.5.0-SNAPSHOT.jar" \
  --conf spark.gluten.loadLibFromJar=true \
  --conf spark.plugins=io.glutenproject.GlutenPlugin \
  --conf spark.gluten.sql.columnar.backend.lib=velox \
  --conf spark.shuffle.manager=org.apache.spark.shuffle.sort.ColumnarShuffleManager \
  --conf spark.memory.offHeap.enabled=true \
  --conf spark.memory.offHeap.size=2G \
  --conf spark.executor.memary=4G \
  --conf spark.executorEnv.LIBHDFS3_CONF="hdfs-site.xml" \
  --files /clusteragent/spark_gluten/conf/hdfs-site.xml \
  --master yarn \
  --keytab /clusteragent/kerberos/keytabs/hadoop.keytab \
  --principal hadoop/slave133
Error info
23/05/30 10:24:45 WARN TaskSetManager: Lost task 0.0 in stage 0.0 (TID 0) (slave132 executor 1): java.lang.RuntimeException: Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: Unable to connect to HDFS: broadtechfed, got error: AccessControlException: Failed to evaluate challenge: GSSAPI error in client while negotiating security context in gss_init_sec_context() in SASL library.  This is most likely due insufficient credentials or malicious interactions..
Retriable: False
Expression: hdfsClient_ != nullptr
Context: Split [file hdfs://broadtechfed/broadtech1root/user/hive/warehouse/inf.db/5g_dpi_ngap_nas_mi/day_id=20230510/hour_id=2023051000/minute_id=202305100000/part-00000-83eb9184-05ba-4dba-8354-008edb500336.c000 0 - 23779] Task gluten task 1
Top-Level Context: Same as context.
Function: Impl
File: ../../velox/connectors/hive/storage_adapters/hdfs/HdfsFileSystem.cpp
Line: 60
Stack trace:
# 0  _ZN8facebook5velox7process10StackTraceC1Ei
# 1  _ZN8facebook5velox14VeloxExceptionC1EPKcmS3_St17basic_string_viewIcSt11char_traitsIcEES7_S7_S7_bNS1_4TypeES7_
# 2  _ZN8facebook5velox6detail14veloxCheckFailINS0_17VeloxRuntimeErrorERKSsEEvRKNS1_18VeloxCheckFailArgsET0_
# 3  _ZN8facebook5velox11filesystems14HdfsFileSystemC2ERKSt10shared_ptrIKNS0_6ConfigEERKNS1_19HdfsServiceEndpointE
# 4  _ZN5folly15basic_once_flagINS_15SharedMutexImplILb0EvSt6atomicNS_24SharedMutexPolicyDefaultEEES2_E14call_once_slowIZNK8facebook5velox11filesystemsUlSt10shared_ptrIKNS8_6ConfigEESt17basic_string_viewIcSt11char_traitsIcEEE_clESD_SH_EUlvE_JEEEvOT_DpOT0_
# 5  _ZNK8facebook5velox11filesystemsUlSt10shared_ptrIKNS0_6ConfigEESt17basic_string_viewIcSt11char_traitsIcEEE_clES5_S9_.isra.0
# 6  _ZNSt17_Function_handlerIFSt10shared_ptrIN8facebook5velox11filesystems10FileSystemEES0_IKNS2_6ConfigEESt17basic_string_viewIcSt11char_traitsIcEEENS3_UlS8_SC_E_EE9_M_invokeERKSt9_Any_dataOS8_OSC_
# 7  _ZN8facebook5velox11filesystems13getFileSystemESt17basic_string_viewIcSt11char_traitsIcEESt10shared_ptrIKNS0_6ConfigEE
# 8  _ZN8facebook5velox19FileHandleGeneratorclERKSs
# 9  _ZN8facebook5velox13CachedFactoryISsNS0_10FileHandleENS0_19FileHandleGeneratorENS0_15FileHandleSizerESt8equal_toISsESt4hashISsEE8generateERKSs
# 10 _ZN8facebook5velox9connector4hive14HiveDataSource8addSplitESt10shared_ptrINS1_14ConnectorSplitEE
# 11 _ZN8facebook5velox4exec9TableScan9getOutputEv
# 12 _ZN8facebook5velox4exec6Driver11runInternalERSt10shared_ptrIS2_ERS3_INS1_13BlockingStateEERS3_INS0_9RowVectorEE
# 13 _ZN8facebook5velox4exec6Driver4nextERSt10shared_ptrINS1_13BlockingStateEE
# 14 _ZN8facebook5velox4exec4Task4nextEPN5folly10SemiFutureINS3_4UnitEEE
# 15 _ZN6gluten24WholeStageResultIterator4nextEv
# 16 Java_io_glutenproject_vectorized_ColumnarBatchOutIterator_nativeHasNext
# 17 0x00007f6e650186c7

	at io.glutenproject.vectorized.ColumnarBatchOutIterator.nativeHasNext(Native Method)
	at io.glutenproject.vectorized.ColumnarBatchOutIterator.hasNextInternal(ColumnarBatchOutIterator.java:45)
	at io.glutenproject.vectorized.GeneralOutIterator.hasNext(GeneralOutIterator.java:37)
	at io.glutenproject.backendsapi.glutendata.GlutenIteratorApi$$anon$2.hasNext(GlutenIteratorApi.scala:240)
	at io.glutenproject.vectorized.CloseableColumnBatchIterator.hasNext(CloseableColumnBatchIterator.scala:41)
	at org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:491)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$getByteArrayRdd$1(SparkPlan.scala:364)
	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2(RDD.scala:890)
	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2$adapted(RDD.scala:890)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:365)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:329)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:136)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:548)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1504)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:551)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

**Comment by zhouyuan** [comment]
@ccat3z can you please do a rebase? there's a minor conflict in main branch

File: ep/build-velox/src/get_velox.sh [hunk 1]
```diff
@@ -2,8 +2,8 @@
 
 set -exu
 
-VELOX_REPO=https://github.com/oap-project/velox.git
-VELOX_BRANCH=main
+VELOX_REPO=https://github.com/ccat3z/velox.git
+VELOX_BRANCH=kerberos
```
Comment (zhouyuan) [inline ep/build-velox/src/get_velox.sh:6]: It looks like this is not required now?

**Comment by zhouyuan** [comment]
Run Gluten Clickhouse CI

File: ep/build-velox/src/get_velox.sh (hunk 1 above)
Comment (ccat3z) [reply ep/build-velox/src/get_velox.sh]: Yes. I have revert it.

File: shims/common/src/main/scala/io/glutenproject/GlutenConfig.scala [hunk 2]
```diff
@@ -358,6 +366,14 @@ object GlutenConfig {
       .filter(_._1.startsWith(backendPrefix))
       .foreach(entry => nativeConfMap.put(entry._1, entry._2))
 
+    // Pass the latest tokens to native
+    nativeConfMap.put(
+      GLUTEN_UGI_TOKENS,
+      UserGroupInformation.getCurrentUser.getTokens.asScala
+        .map(_.encodeToUrlString)
+        .mkString("\0"))
+    nativeConfMap.put(GLUTEN_UGI_USERNAME, UserGroupInformation.getCurrentUser.getUserName)
```
Comment (zhouyuan) [inline shims/common/src/main/scala/io/glutenproject/GlutenConfig.scala:375]: if Kerberos is not enabled, then the username will be `none` ?

Comment (ccat3z) [reply shims/common/src/main/scala/io/glutenproject/GlutenConfig.scala:375]: It will be local linux username.

**Review by zhouyuan — APPROVED** [review]
👍

**Comment by zhouyuan** [comment]
note: will follow up on the doc updates

**Pull request merged**